### PR TITLE
8331920: ubsan: g1CardSetContainers.inline.hpp:266:5: runtime error: index 2 out of bounds for type 'G1CardSetHowl::ContainerPtr [2]' reported

### PR DESCRIPTION
**Pull request by walulyai**
Hi all,

Please review this change to improve access to G1CardSetContainer VLA elements. Instead of  straightforward indexing, we obscure access through a base pointer and offsets to reduce the possibility of UB. 

Testing: Tier 1-3
              - Local testing on Mac with `--enable-ubsan`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8331920: ubsan: g1CardSetContainers.inline.hpp:266:5: runtime error: index 2 out of bounds for type 'G1CardSetHowl::ContainerPtr [2]' reported`

### Issue
 * [JDK-8331920](https://bugs.openjdk.org/browse/JDK-8331920): ubsan: g1CardSetContainers.inline.hpp:266:5: runtime error: index 2 out of bounds for type 'G1CardSetHowl::ContainerPtr [2]' reported (**Bug** - P4)

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Axel Boldt-Christmas](https://openjdk.org/census#aboldtch) (@xmas92 - **Reviewer**) ⚠️ Review applies to [13652dc7](https://git.openjdk.org/jdk/pull/19300/files/13652dc7241d0a40d73891cd081da85aeac4f4c2)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19300/head:pull/19300` \
`$ git checkout pull/19300`

Update a local copy of the PR: \
`$ git checkout pull/19300` \
`$ git pull https://git.openjdk.org/jdk.git pull/19300/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19300`

View PR using the GUI difftool: \
`$ git pr show -t 19300`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19300.diff">https://git.openjdk.org/jdk/pull/19300.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19300#issuecomment-2119934571)

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1CardSetContainers.inline.hpp [hunk 1]
```diff
@@ -257,23 +257,42 @@ inline size_t G1CardSetBitMap::header_size_in_bytes() {
...
 }
 
+inline G1CardSetHowl::ContainerPtr const* G1CardSetHowl::container_addr(EntryCountType index) const {
+  assert(index < _num_entries, "precondition");
+  return buckets() + index;
+}
+
+inline G1CardSetHowl::ContainerPtr* G1CardSetHowl::container_addr(EntryCountType index) {
+  assert(index < _num_entries, "precondition");
+  return const_cast<ContainerPtr*>(buckets() + index);
```
Comment (xmas92) [inline src/hotspot/share/gc/g1/g1CardSetContainers.inline.hpp:267]: While the implementation is very short here, one could chose to implement one in terms of the other.
```suggestion
  return const_cast<ContainerPtr*>(const_cast<const G1CardSetHowl*>(this)->container_addr(index));
```

**Review by tschatzl — APPROVED** [review]
I think this is good.

However I am not sure why ubsan did not complain about the `_data` member of `G1CardSetArray` the same way.

**Review by xmas92 — APPROVED** [review]
The only scary change to me was changing the object layout of `G1CardSetHowl`. But I can find nothing wrong with it, but I also do not understand why it originally was `[2]` instead of `[1]`.

**Review by xmas92 — APPROVED** [review]

**Comment by tschatzl** [comment]
> > However I am not sure why ubsan did not complain about the `_data` member of `G1CardSetArray` the same way.
> 
> It does complain. Should I fix that in this PR? Was planning to fix in follow up.

Follow-up is fine.

File: src/hotspot/share/gc/g1/g1CardSetContainers.hpp [hunk 2]
```diff
@@ -238,23 +238,27 @@ class G1CardSetHowl : public G1CardSetContainer {
...
+  // VLA implementation.
+  ContainerPtr _buckets[1];
+  // Do not add class member variables beyond this point.
 
   // Iterates over the given ContainerPtr with at index in this Howl card set,
   // applying a CardOrRangeVisitor on it.
   template <class CardOrRangeVisitor>
   void iterate_cardset(ContainerPtr const container, uint index, CardOrRangeVisitor& found, G1CardSetConfiguration* config);
 
+  ContainerPtr at(EntryCountType index);
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1CardSetContainers.hpp:250]: Could be made `const`, i.e. 
```suggestion
  ContainerPtr at(EntryCountType index) const;
```

**Review by tschatzl — APPROVED** [review]
Still good.

One of the new methods could be made const, but feel free to ignore (and sorry for noticing a bit late)

**Comment by tschatzl** [comment]
> > > However I am not sure why ubsan did not complain about the `_data` member of `G1CardSetArray` the same way.
> > 
> > 
> > It does complain. Should I fix that in this PR? Was planning to fix in follow up.
> 
> Follow-up is fine.

Actually ubsan did also report that issue about `G1CardSetArray`, it has been added as comment to this issue (https://bugs.openjdk.org/browse/JDK-8331920?focusedId=14671937&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14671937). Still I think it is fine to create an extra CR.

**Comment by walulyai** [comment]
> > > > However I am not sure why ubsan did not complain about the `_data` member of `G1CardSetArray` the same way.
> > > 
> > > 
> > > It does complain. Should I fix that in this PR? Was planning to fix in follow up.
> > 
> > 
> > Follow-up is fine.
> 
> Actually ubsan did also report that issue about `G1CardSetArray`, it has been added as comment to this issue (https://bugs.openjdk.org/browse/JDK-8331920?focusedId=14671937&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14671937). Still I think it is fine to create an extra CR.

Filed under [JDK-8332683](https://bugs.openjdk.org/browse/JDK-8332683)

**Review by tschatzl — APPROVED** [review]

**Pull request closed without merging**